janus_server/src/aggregator.rs

```diff
@@ -933,6 +1010,171 @@ impl VdafOps {
             })
             .await?)
     }
+
+    /// Implements the `/aggregate_share` endpoint for the helper, described in TODO(timg)
+    async fn handle_aggregate_share(
+        &self,
+        datastore: &Datastore,
+        task: &Task,
+        aggregate_share_req: &AggregateShareReq,
+    ) -> Result<AggregateShareResp, Error> {
+        match self {
+            VdafOps::Prio3Aes128Count(_, _) => {
+                Self::handle_aggregate_share_generic::<
+                    <Prio3Aes128Count as Vdaf>::AggregateShare,
+                    FieldError,
+                >(datastore, task, aggregate_share_req)
+                .await
+            }
+            VdafOps::Prio3Aes128Sum(_, _) => {
+                Self::handle_aggregate_share_generic::<
+                    <Prio3Aes128Sum as Vdaf>::AggregateShare,
+                    FieldError,
+                >(datastore, task, aggregate_share_req)
+                .await
+            }
+            VdafOps::Prio3Aes128Histogram(_, _) => {
+                Self::handle_aggregate_share_generic::<
+                    <Prio3Aes128Histogram as Vdaf>::AggregateShare,
+                    FieldError,
+                >(datastore, task, aggregate_share_req)
+                .await
+            }
+
+            #[cfg(test)]
+            VdafOps::Fake(_) => {
+                Self::handle_aggregate_share_generic::<
+                    <fake::Vdaf as Vdaf>::AggregateShare,
+                    Infallible,
+                >(datastore, task, aggregate_share_req)
+                .await
+            }
+        }
+    }
+
+    async fn handle_aggregate_share_generic<A, E>(
+        datastore: &Datastore,
+        task: &Task,
+        aggregate_share_req: &AggregateShareReq,
+    ) -> Result<AggregateShareResp, Error>
+    where
+        A: Aggregatable + Send + Sync,
+        E: std::fmt::Debug,
+        Vec<u8>: for<'a> From<&'a A>,
+        for<'a> A: TryFrom<&'a [u8], Error = E>,
+    {
+        // TODO(timg): What should we do if any of the aggregation jobs handling reports relevant to
+        // the aggregate share request aren't finished yet? Helper could return an error like
+        // "EAGAIN" to leader, or could trust leader not to issue an AggregateShareReq until all of
+        // leader's aggregate jobs are done (which should imply that helper's are done).
+
+        let total_aggregate_share = datastore
+            .run_tx(move |tx| {
+                let task = task.clone();
+                let aggregate_share_req = aggregate_share_req.clone();
+                Box::pin(async move {
```

Comment:
Note that I'm holding the DB transaction open from when we get the aggregations until we increment the consumed lifetime, which I think is necessary to account for multiple concurrent requests to `aggregate_share`.